pkg/manifest/firstboot.go

```diff
@@ -0,0 +1,168 @@
+package manifest
+
+import (
+	"fmt"
+	"html/template"
+	"io/fs"
+	"strings"
+
+	"github.com/osbuild/images/internal/common"
+	"github.com/osbuild/images/pkg/customizations/firstboot"
+	"github.com/osbuild/images/pkg/customizations/fsnode"
+	"github.com/osbuild/images/pkg/osbuild"
+)
+
+// checkName prevents path traversal
+func checkName(str string) error {
+	if str == "" {
+		return fmt.Errorf("name cannot be empty")
+	}
+
+	for _, r := range str {
+		if !(('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || ('0' <= r && r <= '9') || r == '-' || r == '_') {
+			return fmt.Errorf("name can only contain alphanumeric characters, dashes, and underscores")
+		}
+	}
+
+	return nil
+}
+
+var tmplFirstbootAAP = `#!/usr/bin/bash
+curl -s -i --data "host_config_key={{ .HostConfigKey }}" {{ .URL -}}
+`
+
+func renderFirstboot(tmplStr string, data any) (string, error) {
+	tmpl, err := template.New("firstboot-unit").Parse(tmplStr)
+	if err != nil {
+		return "", fmt.Errorf("error parsing firstboot unit template: %w", err)
+	}
+
+	var result strings.Builder
+	err = tmpl.Execute(&result, data)
+	if err != nil {
+		return "", fmt.Errorf("error rendering firstboot unit: %w", err)
+	}
+
+	return result.String(), nil
+}
+
+func firstbootFileNodes(fbo *firstboot.FirstbootOptions) ([]string, []*fsnode.File, *osbuild.SystemdUnitCreateStageOptions, error) {
```

Comment:
It does a bit more than just file nodes right?

Reply:
Yeah refactoring caused this.